File: client-mobile/app/src/main/java/com/example/polyquiz/http/QuestionService.kt
package com.example.polyquiz.http

import com.example.polyquiz.match.domain.Question
import com.google.gson.reflect.TypeToken

object QuestionService : CommunicationService("questions") {
    override val apiService: ApiService = retrofit.create(QuestionsApiService::class.java)

    fun getAllQuestions(onSuccess: (List<Question>) -> Unit, onError: (String) -> Unit) {
        getAll(
            { response -> onSuccess(convertJsonResponseToType(response, object : TypeToken<List<Question>>() {}.type)) },
            onError)
    }

    fun createQuestion(question: Question, onSuccess: (Question) -> Unit, onError: (String) -> Unit) {
        add(
            question,
            { response -> onSuccess(convertJsonResponseToType(response, Question::class.java)) },
            onError
        )
    }
    fun getQuestionById(questionId: String, onSuccess: (Question) -> Unit, onError: (String) -> Unit) {
        getById(
            questionId,
            { question -> onSuccess(convertJsonResponseToType(question, Question::class.java)) },
            onError
        )
    }

    fun deleteQuestion(questionId: String, onSuccess: (Unit) -> Unit, onError: (String) -> Unit) {
        delete(
            questionId,
            onSuccess,
            onError
        )
    }

    fun verifyQuestion(question: Question, onSuccess: (Boolean) -> Unit, onError: (String) -> Unit) {
        add(
            question,
            { response -> onSuccess(convertJsonResponseToType(response, Boolean::class.java)) },
            onError,
            "validate-question"
        )
    }

    fun updateQuestion(modifiedQuestion: Question, onSuccess: (Unit) -> Unit, onError: (String) -> Unit) {
        update(
            modifiedQuestion,
            modifiedQuestion.id,
            onSuccess,
            onError
        )
    }

    interface QuestionsApiService : ApiService {
    }
}

File: client/src/app/services/question/question.service.ts
import { HttpClient, HttpResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { AbstractControl, FormArray, ValidationErrors } from '@angular/forms';
import { MatDialog, MatDialogConfig } from '@angular/material/dialog';
import { QuestionCreationFormComponent } from '@app/components/question-creation-form/question-creation-form.component';
import { ManagementState } from '@app/constants/states';
import { Question } from '@app/interfaces/question';
import { CommunicationService } from '@app/services/communication/communication.service';
import { Observable } from 'rxjs';

export interface DialogManagement {
    modificationState: ManagementState;
}

@Injectable({
    providedIn: 'root',
})
export class QuestionService extends CommunicationService<Question> {
    communcationService: CommunicationService<String>
    constructor(
        public dialog: MatDialog,
        http: HttpClient,

    ) {
        super(http, 'questions');
    }

    getAllQuestions(): Observable<Question[]> {
        return this.getAll();
    }

    generateQuestion(generatedQuestion:string){
        //TODO: AGAIN FIND OUT IF WE CAN MODIFY THE PROMPT SERVER SIDE
        const questionString = {"prompt" : JSON.stringify(generatedQuestion) + "genérer question"}
        console.log(questionString)
        return this.add(questionString,'generate-question');
    }

    createQuestion(question: Question): Observable<HttpResponse<string>> {
        return this.add(question);
    }

    deleteQuestion(questionId: string): Observable<HttpResponse<string>> {
        return this.delete(questionId);
    }

    verifyQuestion(question: Question) {
        return this.add(question, 'validate-question');
    }

    updateQuestion(modifiedQuestion: Question) {
        return this.update(modifiedQuestion, modifiedQuestion.id);
    }

    openCreateQuestionModal(modificationState: ManagementState) {
        const manageConfig: MatDialogConfig<DialogManagement> = {
            data: {
                modificationState,
            },
            height: '50%',
            width: '100%',
        };
        return this.dialog.open(QuestionCreationFormComponent, manageConfig);
    }

    validateChoicesLength(control: AbstractControl): ValidationErrors | null { 
        if (control.get('type')?.value !== 'QCM') return null;
        const choices = control.get('choices') as FormArray;
        let hasCorrect = false;
        let hasIncorrect = false;

        for (let i = 0; i < choices.length; i++) {
            const isCorrect = choices.at(i).get('isCorrect')?.value;

            if (isCorrect) {
                hasCorrect = true;
            } else if (!isCorrect) {
                hasIncorrect = true;
            }
        }

        if (hasCorrect && hasIncorrect) {
            return null;
        } else {
            return { invalidChoicesLength: true };
        }
    }
}
